nomad/secure_variables_endpoint.go

```diff
@@ -367,3 +367,56 @@ func (sv *SecureVariables) decrypt(v *structs.SecureVariable) error {
 	v.EncryptedData = nil
 	return nil
 }
+
+// handleMixedAuthEndpoint is a helper to handle auth on RPC endpoints that can
+// either be called by external clients or by workload identity
+func (sv *SecureVariables) handleMixedAuthEndpoint(args structs.QueryOptions, cap, pathOrPrefix string) error {
+
+	// Perform the initial token resolution.
+	aclObj, err := sv.srv.ResolveToken(args.AuthToken)
+	if err == nil {
+		// Perform our ACL validation. If the object is nil, this means ACLs
+		// are not enabled, otherwise trigger the allowed namespace function.
+		if aclObj != nil {
+			if !aclObj.AllowNsOp(args.RequestNamespace(), cap) {
+				return structs.ErrPermissionDenied
+			}
+		}
+		return nil
+	}
+
+	// Attempt to verify the token as a JWT with a workload
+	// identity claim
+	claim, err := sv.srv.ResolveClaim(args.AuthToken)
+	if err != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	store := sv.srv.fsm.State()
```

Comment:
```suggestion
	store := sv.srv.fsm.State().Snapshot()
```

I think we want to always use a snapshot for readonly cases to (a) ensure a consistent view of state and (b) prohibit changes.

`Server.ResolveToken` _also_ takes a snapshot, so I wonder if they could/should be combined? Memdb snapshots are awfully cheap, so it's probably not worth worrying about. This does mean token resolution may observe multiple states in the course of authenticating a single token, but I can't think of any inconsistencies that could cause.

Reply:
> `Server.ResolveToken` also takes a snapshot, so I wonder if they could/should be combined?

I didn't want to tie allocation lifecycle, which we currently only care about for this RPC endpoint, in with the `ResolveToken` used everywhere else. At least not yet: if we end up making workload identity more broadly useful across RPCs, I think I'd agree but I didn't want to make that call prematurely.